### PR TITLE
issue 689: Added UiBuilder struct

### PR DESCRIPTION
**Pull request by lllShamanlll**
UiBuilder added, allows to create Ui without specifying theme or widgets count.

### CODE REVIEW

**Comment by mitchmindtree** [comment]
Hey @lllShamanlll, thanks for the PR!

I just had a look and have a few notes:

- I think it would be best if the `UiBuilder` only has builder methods - there shouldn't be any need for `set`ters or `get`ters.
- The `UiBuilder`'s `theme` field should be an `Option<Theme>` which defaults to `None`. This will save constructing the `Theme` twice on the occasion that the user specifies some custom `Theme`.
- The `UiBuilder::build` method should take `self` rather than `&self` so that we don't require `clone`ing its contents. This will also allow you to change back to using `Box<Any>` in the `Theme`'s `widget_style` `HashMap` and remove the `Clone` implementation.
- If we're going to use the `UiBuilder` to construct the `Ui`, we should make the `Ui::new` constructor private so that users don't get confused between multiple ways of constructing the `Ui`.

**Comment by lllShamanlll** [comment]
@mitchmindtree  thx for review!
About setters/getters  i've just tried to be consistent with builders in piston.
I'll fix this, but what about getting builder data from Ui::new?
D'u suggest pass all building data as arguments to Ui::new? Doesn't it make Ui::new unscalable?
Or completely remove Ui::new in favor of creating Ui completely on UiBuilder::build side?

**Comment by mitchmindtree** [comment]
No problem!

> About setters/getters i've just tried to be consistent with builders in piston.

I'd prefer we do not include `get`ters and `set`ters as it adds ambiguity to the way that the `UiBuilder` should be used. It should solely be used to construct the `Ui` using the builder pattern.

> I'll fix this, but what about getting builder data from Ui::new?

When specifying the `theme` field of the `Ui` constructor, do this: 
```rust
theme: theme.unwrap_or_else(Theme::default),
```
This will use the `UiBuilder`'s theme if one was specified or constructs the default if it was `None`.

> D'u suggest pass all building data as arguments to Ui::new? Doesn't it make Ui::new unscalable?
> Or completely remove Ui::new in favor of creating Ui completely on UiBuilder::build side?

I'm not exactly sure what you mean by this. I don't mind whether we keep `Ui::new` as a private constructor that the user cannot see (currently [it is still `pub` in your PR](https://github.com/PistonDevelopers/conrod/pull/763/files#diff-46f6f517ae95d07cddf159932baae1a7R176)) or if we inline it into the `UiBuilder::build` method itself. The important thing is that the only way a user should be able to build a `Ui` is using the `UiBuilder::build` method.

File: src/tests/ui.rs [hunk 1]
```diff
@@ -68,7 +69,10 @@ fn to_window_coordinates(xy: Point, ui: &Ui) -> Point {
 
 fn windowless_ui() -> Ui {
     let theme = Theme::default();
-    Ui::new(theme)
+    UiBuilder::new()
+        .theme(theme)
```
Comment (mitchmindtree) [inline src/tests/ui.rs]: Remember this is no longer necessary as `Theme::default()` is now the default.

File: src/ui.rs [hunk 2]
```diff
@@ -88,36 +97,62 @@ pub struct UiCell<'a> {
...
-                    depth_order: graph::DepthOrder,
-                    updated_widgets: std::collections::HashSet<NodeIndex>) -> Self
-    {
         let window = widget_graph.add_placeholder();
         let prev_updated_widgets = updated_widgets.clone();
-        Ui {
+        Ok(Ui {
             widget_graph: widget_graph,
-            theme: theme,
+            theme: builder.maybe_theme.unwrap_or(Theme::default()),
```
Comment (mitchmindtree) [inline src/ui.rs]: This line always constructs the `Theme::default()`, even if `maybe_theme` is `Some`.

Instead, a more efficient alternative here would be: 
```rust
theme: builder.maybe_theme.unwrap_or_else(Theme::default),
```

Here, the function `Theme::default` is instead only called if `maybe_theme` is `None`.

File: src/ui.rs [hunk 3]
```diff
@@ -88,36 +97,62 @@ pub struct UiCell<'a> {
 /// buffer. Otherwise if we don't draw into each buffer, we will probably be subject to flickering.
 pub const SAFE_REDRAW_COUNT: u8 = 3;
 
+impl UiBuilder {
+    /// Create **Ui** builder with defaults.
+    ///
+    /// - theme: Theme::default()
```
Comment (mitchmindtree) [inline src/ui.rs]: Should this be changed to `maybe_theme`?

File: src/ui.rs [hunk 4]
```diff
@@ -88,36 +97,62 @@ pub struct UiCell<'a> {
...
+            maybe_theme: None,
+            maybe_widgets_capacity: None
+        }
+    }
 
-impl Ui {
+    /// Sets the theme of built **Ui**.
+    ///
+    /// This method moves the current builder,
+    /// unlike [`set_theme()`](#method.set_theme),
```
Comment (mitchmindtree) [inline src/ui.rs]: There is no longer any `set_theme` method, perhaps these doc comments should be updated?

**Pull request closed without merging**